core/src/transport/mod.rs

```diff
@@ -54,6 +56,45 @@ pub use self::memory::connector;
 pub use self::muxed::MuxedTransport;
 pub use self::upgrade::UpgradedNode;
 
+
+#[derive(Debug)]
+pub enum TransportError {
+    ListenNotSupported(Multiaddr),
+    ListenFailed(Multiaddr, IoError),
+    DialNotSupported(Multiaddr),
+    DialingFailed(Multiaddr, IoError),
+}
+
+impl fmt::Display for TransportError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            TransportError::ListenNotSupported(ref addr) =>
+                f.write_str(&format!("Listening is not support for Multiaddress {}", addr)),
+            TransportError::ListenFailed(ref addr, ref io_err) =>
+                f.write_str(&format!("Listening on Multiaddress {} failed: {}", addr, io_err)),
+            TransportError::DialNotSupported(ref addr) =>
+                f.write_str(&format!("Dialing is not support for Multiaddress {}", addr)),
+            TransportError::DialingFailed(ref addr, ref io_err) =>
+                f.write_str(&format!("Dialing for Multiaddress {} failed: {}", addr, io_err)),
+        }
+    }
+}
+
+
+impl StdError for TransportError {
+    fn description(&self) -> &str {
+        match *self {
+            TransportError::ListenNotSupported(_) =>  "Listening is not support",
+            TransportError::ListenFailed(_, _) => "Listening failed",
+            TransportError::DialNotSupported(_) => "Dialing is not support",
+            TransportError::DialingFailed(_, _) => "Dialing failed",
+
+        }
+    }
+}
```

Comment:
`std::error::Error::description` is "soft-deprecated", no need to implement it. OTOH, it may be helpful to add an implementation for `cause` as some error cases wrap underlying `io::Error`s.